Add new features

iloveicedgreentea opened the pull request:
Hey, I am the maintainer for another JVC integration

https://github.com/iloveicedgreentea/jvc_projector_python

https://github.com/iloveicedgreentea/jvc_homeassistant

I have archived it since yours is in core already. I plan on extending the core version to support new sensors and other entities that my custom component had. As a precursor, I wanted to add some features to the library

Summary of changes
1) Added a bunch of new commands
2) Update state function to be modular
3) Fixed some misc bugs
4) Lowered throttling to speed up state updates (maybe this should be configurable)
5) added helper functions to transform values to human readable ones like version, model, mac, etc

I updated some tests and made sure all of these worked with my NZ7. I like how you set up the formatter with regex for parsing commands, that is very slick.

Please let me know what you think

iloveicedgreentea commented:
I will add the new handshake procedure for new models tomorrow as well

iloveicedgreentea commented:
Added the new auth protocol and a new way to send commands. It uses a send_command function that can send CMD consts. This will allow home assistant to send commands like set the laser_power to high (added a test for that)

iloveicedgreentea commented:
Heres some output examples

```
{'model': 'ILAFPJ-B5A3', 'mac': 'E0DADC0A269A'}

Current state:
model: NZ7 <- model code lookup 
version: 3.0.0 <- version is semver now
mac: E0:DA:DC:0A:26:9A <- format mac
power: on
input: hdmi2
source: signal
picture_mode: user1
low_latency: off
laser_power: high
anamorphic: off
installation_mode: mode5
hdr: none
hdmi_input_level: standard
hdmi_color_space: auto
color_profile: auto
graphics_mode: high-res
color_space: rgb
eshift: off
clear_motion_drive: off
motion_enhance: low
laser_value: 100 <- this is normalized to 0-100
laser_time: 2195 <- hours
laser_dimming: auto3
hdr_processing: None
hdr_content_type: auto
```

I added a functional test so you can check it out as well

iloveicedgreentea commented:
Any chance this could get reviewed?

jmery commented:
@SteveEasley Would it be possible to get this reviewed and merged?  This add-on is broken for all of the new JVC models due to a password format change.  @iloveicedgreentea has that fixed above so it would be great to implement.

iloveicedgreentea commented:
It’s been a while so I might just have to fork this and make a PR to home assistant to use the fork

SteveEasley commented:
Hey sorry, just noticed this request after ironically just pushing up some changes to fix the auth issue. Not sure why I didn't get a git notification of the PR until I pushed up my change.

@iloveicedgreentea I notice our auth code is a little different. I think your implementation might be including the `_` in the sha256 encode, which is a mistake? I also added a test to verify the functionality (I dont have a newer projector to test though).

The const changes look like a good thing. But I do want to take a day or so to test the other new commands on my own projector. I have some concerns about the fast refreshes and bogging down the projector with the new traffic. But might be ok.

How about we split this up into two HA PRs? I will get my auth changes pushed through, and lets work on the new commands as a separate (but concurrent) change. Feel free to give feedback on my changes though (https://github.com/SteveEasley/pyjvcprojector/commit/05d8e86f7294983121be9b1705fb76434f2ce61a)

iloveicedgreentea commented:
> Hey sorry, just noticed this request after ironically just pushing up some changes to fix the auth issue. Not sure why I didn't get a git notification of the PR until I pushed up my change.
> 
> @iloveicedgreentea I notice our auth code is a little different. I think your implementation might be including the `_` in the sha256 encode, which is a mistake? I also added a test to verify the functionality (I dont have a newer projector to test though).
> 
> The const changes look like a good thing. But I do want to take a day or so to test the other new commands on my own projector. I have some concerns about the fast refreshes and bogging down the projector with the new traffic. But might be ok.
> 
> How about we split this up into two HA PRs? I will get my auth changes pushed through, and lets work on the new commands as a separate (but concurrent) change. Feel free to give feedback on my changes though ([05d8e86](https://github.com/SteveEasley/pyjvcprojector/commit/05d8e86f7294983121be9b1705fb76434f2ce61a))

Sure sounds good I copied the implementation from my library which worked for users but could have added an extra underscore there. I don’t think the underscore is supposed to be included.  I can rebase without the auth changes when it’s merged

SteveEasley commented:
The auth change is already merge to main here (will be good to have eyes for PRs in the future though!). So feel free to rebase when you are ready. I am starting to dig into your changes now, with my projector.

SteveEasley commented:
@jmery You should actually be able work around the issue until we get this into HA by doing the password encode yourself. 

In fact this might be a good test of our future encoding logic, since you would be manually doing exactly what's being done in the library. Also you might be able to test an edge case I am worried about (short passwords).

1. Add a 10+ character to your projector if it doesn't have one already (e.g. `1234567890`).
2. Now SHA256 encode that password + the string `JVCKWPJ` into an online SHA256 calculator https://www.google.com/search?q=sha256+encode+online. (e.g. `1234567890JVCKWPJ` is encoded as `cc3054cf7c701a32a883bf45a7b19cbf7f62db778208b3c058e0c6dde886996f`).
3. Use that encoded password as your password in the JVC Projector integration of HomeAssistant. I think you might have to remove the integration first, since there is no way to change or edit the password afterward.

If that works, would you mind testing a password less than 10 characters? The current library null pads the password to at least 10 characters and I want to make sure that is still a valid thing to do with these new encoded passwords.

Thanks!

SteveEasley commented on jvcprojector/projector.py:
```diff
@@ -101,50 +100,140 @@ async def disconnect(self) -> None:
+
+        model = JvcCommand(const.CMD_MODEL, True)
+        mac = JvcCommand(const.CMD_LAN_SETUP_MAC_ADDRESS, True)
+        version = JvcCommand(const.CMD_VERSION, True)
```
CMD_VERSION (IFSV) is timing out 100% of the time when my projector is not in ON (in STANDBY or COOLING). I tried bumping up the sleep between commands but made no difference.

```
2024-09-23 11:01:32.103 DEBUG (MainThread) [jvcprojector.device] Connecting to 192.168.0.30
2024-09-23 11:01:32.116 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJ_OK'
2024-09-23 11:01:32.118 DEBUG (MainThread) [jvcprojector.device] Handshake sending 'PJREQ'
2024-09-23 11:01:32.120 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJACK'
2024-09-23 11:01:32.120 DEBUG (MainThread) [jvcprojector.device] Sending op ' (b'!\x89\x01\x00\x00\n')'
2024-09-23 11:01:32.121 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01\x00\x00\n'
2024-09-23 11:01:32.322 DEBUG (MainThread) [jvcprojector.device] Sending ref 'MD (b'?\x89\x01MD\n')'
2024-09-23 11:01:32.325 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01MD\n'
2024-09-23 11:01:32.425 DEBUG (MainThread) [jvcprojector.device] Received ref b'ILAFPJ -- B2A2' (b'@\x89\x01MDILAFPJ -- B2A2\n')
2024-09-23 11:01:32.626 DEBUG (MainThread) [jvcprojector.device] Sending ref 'LSMA (b'?\x89\x01LSMA\n')'
2024-09-23 11:01:32.633 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01LS\n'
2024-09-23 11:01:32.733 DEBUG (MainThread) [jvcprojector.device] Received ref b'E0DADC0A1562' (b'@\x89\x01LSE0DADC0A1562\n')
2024-09-23 11:01:32.934 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IFSV (b'?\x89\x01IFSV\n')'
(long delay)
2024-09-23 11:01:47.937 WARNING (MainThread) [jvcprojector.device] Response timeout for 'IFSV'
2024-09-23 11:01:48.139 DEBUG (MainThread) [jvcprojector.device] Disconnected
2024-09-23 11:01:48.139 ERROR (MainThread) [homeassistant.components.jvc_projector.coordinator] Unexpected error fetching JVC Projector data
Traceback (most recent call last):
  File "/Volumes/code/python/home-assistant/homeassistant/helpers/update_coordinator.py", line 354, in _async_refresh
    self.data = await self._async_update_data()
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/code/python/home-assistant/homeassistant/components/jvc_projector/coordinator.py", line 47, in _async_update_data
    state = await self.device.get_state()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/code/python/iloveicedgreentea/pyjvcprojector/jvcprojector/projector.py", line 134, in get_state
    self._dict[const.KEY_VERSION] = self.process_version(self._version)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Volumes/code/python/iloveicedgreentea/pyjvcprojector/jvcprojector/projector.py", line 213, in process_version
    version = version.removesuffix("PJ")
              ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'removesuffix'
2024-09-23 11:01:48.141 DEBUG (MainThread) [homeassistant.components.jvc_projector.coordinator] Finished fetching JVC Projector data in 0.002 seconds (success: False)
```

iloveicedgreentea replied:
added a check for power state

SteveEasley commented on jvcprojector/projector.py:
```diff
@@ -101,50 +100,140 @@ async def disconnect(self) -> None:
+            if self._dict.get("hdr") != const.HDR_CONTENT_SDR:
+                await send_and_update(
+                    {
+                        const.KEY_HDR_PROCESSING: const.CMD_PICTURE_MODE_HDR_PROCESSING,
```
CMD_PICTURE_MODE_HDR_PROCESSING (PMHP) is timing out on my RS2000/NX7 100% of the time when there is no HDR source. 

```
2024-09-23 11:07:58.584 DEBUG (MainThread) [jvcprojector.device] Connecting to 192.168.0.30
2024-09-23 11:07:58.596 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJ_OK'
2024-09-23 11:07:58.597 DEBUG (MainThread) [jvcprojector.device] Handshake sending 'PJREQ'
2024-09-23 11:07:58.598 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJACK'
2024-09-23 11:07:58.598 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PW (b'?\x89\x01PW\n')'
2024-09-23 11:07:58.601 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PW\n'
2024-09-23 11:07:58.601 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PW1\n')
2024-09-23 11:07:58.803 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IP (b'?\x89\x01IP\n')'
2024-09-23 11:07:58.805 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IP\n'
2024-09-23 11:07:58.906 DEBUG (MainThread) [jvcprojector.device] Received ref b'6' (b'@\x89\x01IP6\n')
2024-09-23 11:07:59.107 DEBUG (MainThread) [jvcprojector.device] Sending ref 'SC (b'?\x89\x01SC\n')'
2024-09-23 11:07:59.111 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01SC\n'
2024-09-23 11:07:59.213 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01SC0\n')
2024-09-23 11:07:59.414 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPM (b'?\x89\x01PMPM\n')'
2024-09-23 11:07:59.417 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:07:59.521 DEBUG (MainThread) [jvcprojector.device] Received ref b'03' (b'@\x89\x01PM03\n')
2024-09-23 11:07:59.723 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLL (b'?\x89\x01PMLL\n')'
2024-09-23 11:07:59.726 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:07:59.829 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:08:00.030 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLP (b'?\x89\x01PMLP\n')'
2024-09-23 11:08:00.033 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:08:00.133 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:08:00.334 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INVS (b'?\x89\x01INVS\n')'
2024-09-23 11:08:00.338 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:08:00.437 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IN0\n')
2024-09-23 11:08:00.638 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INML (b'?\x89\x01INML\n')'
2024-09-23 11:08:00.642 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:08:00.742 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IN1\n')
2024-09-23 11:08:00.944 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMHP (b'?\x89\x01PMHP\n')'
>>>>>>>> 2024-09-23 11:08:15.945 WARNING (MainThread) [jvcprojector.device] Response timeout for 'PMHP'
2024-09-23 11:08:16.147 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMCT (b'?\x89\x01PMCT\n')'
2024-09-23 11:08:16.153 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:08:16.253 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:08:16.454 DEBUG (MainThread) [homeassistant.components.jvc_projector.coordinator] Finished fetching JVC Projector data in 17.870 seconds (success: True)
2024-09-23 11:08:18.455 DEBUG (MainThread) [jvcprojector.device] Disconnected
```

Then when an HDR signal becomes available it doesnt timeout, but gives the error `Index 3 out of range for command PMHP`. Annoyingly `3` is not documented in my V3.0 command reference/

```
2024-09-23 11:18:49.603 DEBUG (MainThread) [jvcprojector.device] Connecting to 192.168.0.30
2024-09-23 11:18:49.617 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJ_OK'
2024-09-23 11:18:49.617 DEBUG (MainThread) [jvcprojector.device] Handshake sending 'PJREQ'
2024-09-23 11:18:49.619 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJACK'
2024-09-23 11:18:49.619 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PW (b'?\x89\x01PW\n')'
2024-09-23 11:18:49.621 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PW\n'
2024-09-23 11:18:49.621 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PW1\n')
2024-09-23 11:18:49.823 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IP (b'?\x89\x01IP\n')'
2024-09-23 11:18:49.825 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IP\n'
2024-09-23 11:18:49.925 DEBUG (MainThread) [jvcprojector.device] Received ref b'6' (b'@\x89\x01IP6\n')
2024-09-23 11:18:50.126 DEBUG (MainThread) [jvcprojector.device] Sending ref 'SC (b'?\x89\x01SC\n')'
2024-09-23 11:18:50.132 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01SC\n'
2024-09-23 11:18:50.232 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01SC1\n')
2024-09-23 11:18:50.433 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPM (b'?\x89\x01PMPM\n')'
2024-09-23 11:18:50.439 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:50.539 DEBUG (MainThread) [jvcprojector.device] Received ref b'0B' (b'@\x89\x01PM0B\n')
2024-09-23 11:18:50.741 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLL (b'?\x89\x01PMLL\n')'
2024-09-23 11:18:50.747 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:50.847 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:18:51.048 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLP (b'?\x89\x01PMLP\n')'
2024-09-23 11:18:51.053 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:51.153 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:18:51.354 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INVS (b'?\x89\x01INVS\n')'
2024-09-23 11:18:51.361 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:18:51.461 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IN0\n')
2024-09-23 11:18:51.663 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INML (b'?\x89\x01INML\n')'
2024-09-23 11:18:51.669 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:18:51.769 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IN1\n')
2024-09-23 11:18:51.971 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IFHR (b'?\x89\x01IFHR\n')'
2024-09-23 11:18:51.978 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IF\n'
2024-09-23 11:18:52.078 DEBUG (MainThread) [jvcprojector.device] Received ref b'2' (b'@\x89\x01IF2\n')
2024-09-23 11:18:52.279 DEBUG (MainThread) [jvcprojector.device] Sending ref 'ISIL (b'?\x89\x01ISIL\n')'
2024-09-23 11:18:52.286 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IS\n'
2024-09-23 11:18:52.386 DEBUG (MainThread) [jvcprojector.device] Received ref b'3' (b'@\x89\x01IS3\n')
2024-09-23 11:18:52.587 DEBUG (MainThread) [jvcprojector.device] Sending ref 'ISHS (b'?\x89\x01ISHS\n')'
2024-09-23 11:18:52.593 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IS\n'
2024-09-23 11:18:52.693 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IS0\n')
2024-09-23 11:18:52.894 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPR (b'?\x89\x01PMPR\n')'
2024-09-23 11:18:52.900 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.000 DEBUG (MainThread) [jvcprojector.device] Received ref b'24' (b'@\x89\x01PM24\n')
2024-09-23 11:18:53.201 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMGM (b'?\x89\x01PMGM\n')'
2024-09-23 11:18:53.207 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.306 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:18:53.508 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IFXV (b'?\x89\x01IFXV\n')'
2024-09-23 11:18:53.511 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IF\n'
2024-09-23 11:18:53.614 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IF1\n')
2024-09-23 11:18:53.815 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMHP (b'?\x89\x01PMHP\n')'
2024-09-23 11:18:53.818 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.922 DEBUG (MainThread) [jvcprojector.device] Received ref b'3' (b'@\x89\x01PM3\n')
2024-09-23 11:18:54.123 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMCT (b'?\x89\x01PMCT\n')'
2024-09-23 11:18:54.126 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:54.226 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
>>>>>> 2024-09-23 11:18:54.427 WARNING (MainThread) [jvcprojector.command] Index 3 out of range for command PMHP
2024-09-23 11:18:54.428 DEBUG (MainThread) [homeassistant.components.jvc_projector.coordinator] Finished fetching JVC Projector data in 4.824 seconds (success: True)
^C2024-09-23 11:18:55.093 DEBUG (MainThread) [jvcprojector.device] Disconnected
```

SteveEasley replied:
Ah at least for `Index 3 out of range error`, after looking at your lib I see I am missing the 0th index in my value definition like you have. 

What do you use as a reference for these values? The latest reference I know of is http://pro.jvc.com/pro/attributes/PRESENT/Manual/External%20Command%20Spec%20for%20D-ILA%20projector_V3.0.pdf. But that one has no 0th element for `hdr10_plus` like you have.

SteveEasley replied:
Ok after further digging, line 189 below is whats allowing the call to PMHP to erroniously happen, because in my case the HDR value is `none(F)`, not `SDR(0)`.

```python
if self._dict.get("hdr") != const.HDR_CONTENT_SDR:
```

iloveicedgreentea replied:
updated to check for none as well

SteveEasley commented on requirements_test.txt:
```diff
@@ -1,6 +1,6 @@
 aioconsole==0.5.1
 black==22.12.0
-flake8=6.0.0
+flake8==6.0.0
```
Today in main I replaced these lint/format tools with ruff (same as HA uses), and added a pre-commit hook for it.

iloveicedgreentea commented:
Thanks I will take a look at these this week

jmery commented:
> @jmery You should actually be able work around the issue until we get this into HA by doing the password encode yourself.
> 
> In fact this might be a good test of our future encoding logic, since you would be manually doing exactly what's being done in the library.
> 
> 1. Add a password to your projector if it doesn't have one already (e.g. `1234567890`).
> 2. Now SHA256 encode that password + the string `JVCKWPJ` into an online SHA256 calculator https://www.google.com/search?q=sha256+encode+online. (e.g. `1234567890JVCKWPJ` is encoded as `cc3054cf7c701a32a883bf45a7b19cbf7f62db778208b3c058e0c6dde886996f`).
> 3. Use that encoded password as your password in the JVC Projector integration of HomeAssistant. I think you might have to remove the integration first, since there is no way to change or edit the password afterward.
> 
> Thanks!

This doesn't appear to work.  I reset the PJ's network password.  Adding the integration via the HA UI constantly returns "Invalid Authentication".

SteveEasley commented:
Hmm that’s concerning since that’s the same process we would use in the code. 

Would you mind trying with a test password and share it and the sha256 value you put into the HA integration with me?

jmery commented:
> Hmm that’s concerning since that’s the same process we would use in the code.
> 
> Would you mind trying with a test password and share it and the sha256 value you put into the HA integration with me?
> 
> And/or try it the exact values I gave above. 1234567890 for the projector password and cc3054cf7c701a32a883bf45a7b19cbf7f62db778208b3c058e0c6dde886996f for the HA integration password.

I reset it to your exact example from the original password.  I don't see any instructions from JVC requiring any kind of reboot or anything for the password to be effective.

SteveEasley commented:
Ok figured this out. This was actually an unrelated bug thats always been there. Passwords over 10 characters were getting truncated to 10 characters, so it wont except the long SHA :( I submitted a fix for this here https://github.com/home-assistant/core/pull/126581. But by the time thats fixed, but SHA256 PR (https://github.com/home-assistant/core/pull/126453) should be live at the same time, so should be able to use your regular password.

Thanks for taking the time to try it, squashed a bug at least!

jmery commented:
> Ok figured this out. This was actually an unrelated bug thats always been there. Passwords over 10 characters were getting truncated to 10 characters, so it wont except the long SHA :( I submitted a fix for this here [home-assistant/core#126581](https://github.com/home-assistant/core/pull/126581). But by the time thats fixed, the SHA256 PR ([home-assistant/core#126453](https://github.com/home-assistant/core/pull/126453)) should be live at the same time, so should be able to use your regular password.
> 
> Thanks for taking the time to try it, squashed a bug at least!

2024.9.3 shipped overnight.  I just installed and can confirm the issue is fixed.  Thanks so much for the quick fix and happy to have helped a little :-)

iloveicedgreentea commented:
I rebased to main, used your connect code, and fixed the rest of the comments. Please let me know what you think

SteveEasley commented:
Cool will look at this over the next couple of days.

SteveEasley commented on jvcprojector/projector.py:
```diff
@@ -101,50 +104,147 @@ async def disconnect(self) -> None:
+        if self._dict.get(const.KEY_POWER) == const.ON:
+            await send_and_update(
+                {
+                    const.KEY_INPUT: const.CMD_INPUT,
```
Guess the HA integration can be reworked some, but it relies on the `input` key being in the response at all times (even it set to NO_SIGNAL). https://github.com/home-assistant/core/blob/dev/homeassistant/components/jvc_projector/select.py#L32

This was erroring out for me when I tested with your PR in HA.

iloveicedgreentea replied:
the HA changes will require a PR for a version bump so IMO this wont break anything as I plan to add these features to the integration at the same time

iloveicedgreentea commented on jvcprojector/projector.py:
```diff
@@ -101,50 +104,147 @@ async def disconnect(self) -> None:
+        # store to reference in state
         self._model = model.response
         self._mac = mac.response
+        self._version = version.response
```
pushed a change lmk if thats what you had in mind

```if version.response is None:
            version.response = "(unknown)"
```
otherwise it will default to none

SteveEasley replied:
The issue is `version` is not defined in line 128 if the projector is off because in line 112 it's skipped. I think its ok to just always get the version from the projector.

SteveEasley replied:
Also the string `(unknown)` will fail the `process_version()` parse further down.

iloveicedgreentea commented on jvcprojector/projector.py:
```diff
@@ -101,50 +104,147 @@ async def disconnect(self) -> None:
+        model = JvcCommand(const.CMD_MODEL, True)
+        # ensure PJ is fully on first
+        if await self.is_on():
+            mac = JvcCommand(const.CMD_LAN_SETUP_MAC_ADDRESS, True)
```
okay I thought it was not available until on

iloveicedgreentea commented on jvcprojector/projector.py:
```diff
@@ -180,3 +280,58 @@ async def _send(self, cmds: list[JvcCommand]) -> list[str | None]:
+    def _invert_dict(d: dict[str, str]) -> dict[str, str]:
+        return {v: k for k, v in d.items()}
+
+    def _build_command_map(self) -> dict[str, dict[str, Any]]:
```
yep done

iloveicedgreentea commented:
Pushed changes to address comments. Tests pass

Heres what the output looks like

{'model': 'ILAFPJ-B5A3', 'mac': 'E0DADC0A269A'}
Current state:
model: NZ7
version: 3.0.0
mac: E0:DA:DC:0A:26:9A
power: on
input: hdmi2
source: signal
picture_mode: user1
low_latency: off
laser_power: high
anamorphic: off
installation_mode: mode5
hdr: none
hdmi_input_level: standard
hdmi_color_space: auto
color_profile: auto
graphics_mode: high-res
color_space: rgb
eshift: off
clear_motion_drive: off
motion_enhance: low
laser_value: 100
laser_time: 2246
laser_dimming: auto3
hdr_processing: None
hdr_content_type: auto

iloveicedgreentea commented on jvcprojector/projector.py:
```diff
@@ -101,50 +100,140 @@ async def disconnect(self) -> None:
 
         if mac.response is None:
-            raise JvcProjectorError("Mac address not available")
+            mac.response = "(unknown)"
```
yeah that should be critical then I will make it an error

iloveicedgreentea commented:
pushed a fix to improve sending info commands and fix version processing for unknown string

SteveEasley commented:
fyi looking good. I am assuming we will leave this open until you are done on the HA side and ready for a PR on that side.

iloveicedgreentea commented:
> fyi looking good. I am assuming we will leave this open until you are done on the HA side and ready for a PR on that side.

Yeah I started working on that in core. I will try to get it plugged in soonish

iloveicedgreentea commented:
@SteveEasley I made a draft PR for HA core here https://github.com/home-assistant/core/pull/131320/files

I added sending commands and the new sensors. I need to do more testing but at first glance it all works. I am installing the lib from git (pulling changes into my core devcontainer) with

```
pip install  /workspaces/pyjvcprojector/
```

Makes it easy to iterate

LMK what you think. I will probably disable some sensors by default because the core people will complain about it. I also want to make it easier to see what commands can be sent like "laser_power, high".

I also updated the tests.

iloveicedgreentea commented:
I pushed changes to automate generating commands and for HA I added more selects and improved sensors and all that. Will do more testing tomorrow. 99% code coverage too

```
---------------------------------------------------------------------------------------
homeassistant/components/jvc_projector/__init__.py           31      1    97%   50
homeassistant/components/jvc_projector/binary_sensor.py      29      0   100%
homeassistant/components/jvc_projector/config_flow.py        62      0   100%
homeassistant/components/jvc_projector/const.py               5      0   100%
homeassistant/components/jvc_projector/coordinator.py        31      0   100%
homeassistant/components/jvc_projector/entity.py             17      0   100%
homeassistant/components/jvc_projector/remote.py             44      1    98%   71
homeassistant/components/jvc_projector/select.py             33      0   100%
homeassistant/components/jvc_projector/sensor.py             25      0   100%
---------------------------------------------------------------------------------------
TOTAL                                                       277      2    99%
```

iloveicedgreentea commented:
<img width="677" alt="image" src="https://github.com/user-attachments/assets/3ea09337-f64f-4684-8120-df8d8a1d42c3">

Updated model and version display as well. I tested the selects, it all seems to work. Please test and let me know. You will probably have to edit your manifest to have this.



```
{
  "domain": "jvc_projector",
  "name": "JVC Projector",
  "codeowners": ["@SteveEasley", "@msavazzi"],
  "config_flow": true,
  "documentation": "https://www.home-assistant.io/integrations/jvc_projector",
  "integration_type": "device",
  "iot_class": "local_polling",
  "loggers": ["jvcprojector"],
  "requirements": ["/workspaces/pyjvcprojector/"]
}

```

iloveicedgreentea commented:
@SteveEasley ive been running this in my HA instance and its been good the only issues I see are occasional timeouts

<img width="531" alt="image" src="https://github.com/user-attachments/assets/7cac7909-ab3a-48d0-a42f-d46420e653c5">


I suspect its an issue of trying to read an empty buffer which suggests a syncing issue but not sure. Not sure if this ever happened before my changes. In my custom component I implemented a PriorityQueue which allowed user commands to have immediate access while refresh commands took a back seat. This gave you strict command ordering, priority, and a queue system. let me know if you are interested in that I can implement it

SteveEasley commented:
Hey, sorry for the delay. Should be diving into this by the end of the week.

Sdub76 commented:
> Hey, sorry for the delay. Should be diving into this by the end of the week.

Any update?  I'm waiting for this to merge before switching to the official JVC integration, since I use installation mode in my automations.  Was hoping to make the update over the holidays.  Thanks!

SteveEasley commented:
> Any update? I'm waiting for this to merge before switching to the official JVC integration, since I use installation mode in my automations. Was hoping to make the update over the holidays. Thanks!

Yep I provided some feedback on the HA side in https://github.com/home-assistant/core/pull/131320 last weekend. It was working for me for the most part.

iloveicedgreentea commented:
I pushed some changes here and to core. I do notice that user commands could get ignored during an update which is what I think causes the timeout warnings. Since this PR is already huge it would make sense to implement a priority queue after this which would also have the added benefit of making user commands instant (plus all the other stuff like order guarantees)

iloveicedgreentea commented:
> > Hey, sorry for the delay. Should be diving into this by the end of the week.
> 
> Any update? I'm waiting for this to merge before switching to the official JVC integration, since I use installation mode in my automations. Was hoping to make the update over the holidays. Thanks!

If you know what you are doing you can install my PR in core as a custom component and upload this PR to pypy (or I suppose just reference a directory instead)

SteveEasley commented:
I did added some changes for consideration to https://github.com/iloveicedgreentea/core/pull/1

crazyshakespeare13 commented:
Probably a very stupid question but I am new....I was apply to get the JVC projector integration to install and see my projector but I am clueless as to how to control the projector from home assistant now that its in there...Any help would be appreciated

Upke commented:
I’ve been using this component as a custom component for the last two months. It has been working great! I greatly value the added fearures. There is the occasional error in the log, but nothing breaking. An example of an error I receive:

Logger: jvcprojector.device
Bron: custom_components/jvc_projector/select.py:32
integratie: JVC Projector
Eerst voorgekomen: 7 februari 2025 om 21:26:00 (1 gebeurtenissen)
Laatst gelogd: 7 februari 2025 om 21:26:00

Response timeout for 'INML0'

I do had to map out/discover the different sensors and controles due to the not being labeled. Perhaps something with my install, or with the component itself? 

I was wondering, what actions are still needed to merge this?

rcdailey commented:
Seems like progress has slowed here. What's left holding up a merge? Is there any action items that the community can assist with? Thanks for the effort!

SteveEasley commented:
Looks like HA closed the PR on that side due to the size (unfortunate, but common). iloveicedgreentea mentioned he wont have time to manage that process. I can pick it up from here and start splitting things out. Thanks for all the work iloveicedgreentea!

momala454 commented:
hello @SteveEasley did you stark working on it ? Thanks
